### PR TITLE
InstallPlan fixes and misc housekeeping

### PR DESCRIPTION
**Pull request by dcoutts**
This subsumes PR #3888 and fixes issue #3889.

We worked out what the actual problem with the InstallPlan processing invariant failures were. The invariant was wrong. See the discussion in #3889 for details.

This PR fixes the invariant, and adds a couple other parts to the invariant that we think are true. Also add commentary on the invariant and do something similar to what PR #3888 does (but a bit more lightweight) so we get more informative assertion violations in future.

Finally, do a bit of housekeeping, rename some things and export a few more utils that are going to be useful in other upcoming PRs.

### CODE REVIEW

File: cabal-install/Distribution/Client/InstallPlan.hs [hunk 1]
```diff
@@ -596,9 +596,9 @@ processingInvariant plan (Processing processingSet completedSet failedSet) =
  && all (isJust . flip Graph.lookup (planIndex plan)) (Set.toList failedSet)
  && noIntersection processingSet completedSet
  && noIntersection processingSet failedSet
+ --   intersection processingClosure failedSet is quite possible however
```
Comment (ezyang) [inline cabal-install/Distribution/Client/InstallPlan.hs]: Probably didn't want this line (you delete it later anyway).

**Comment by ezyang** [comment]
LGTM. But build failure is REAL. Do a local build `-w ghc-8.0 -O0` and squash the warnings ;) (I wish `--ghc-options=-Werror` worked but it tries to rebuild everything.)

File: cabal-install/Distribution/Client/InstallPlan.hs (hunk 1 above)
Comment (dcoutts) [inline cabal-install/Distribution/Client/InstallPlan.hs]: Well, the comment makes sense in the context of the patch, but it does get replaced by a better comment later. So it does no harm.

**Comment by dcoutts** [comment]
@ezyang Ta. Done.

Also added a related assertion failure fix.

**Comment by ezyang** [comment]
`git rebase --ignore-date` I think?

**Pull request merged**